ignition/include/gympp/gazebo/IgnitionRobot.h

```diff
@@ -71,6 +74,8 @@ class gympp::gazebo::IgnitionRobot : public gympp::Robot
 
     bool setJointPosition(const JointName& jointName, const double jointPosition) override;
     bool setJointVelocity(const JointName& jointName, const double jointVelocity) override;
+    bool setJointControlMode(const JointName& jointName,
+                             const JointControlMode controlMode) override;
```

Comment:
Not a big problem for the simulation runtimes, but for real robot runtimes it may be useful to be able to set also a first reference when you switch control mode. If there is the possibility of setting the joint position setpoints synchronously with the joint control mode, then there is  no problem.

Reply:
Excellent catch @traversaro. Right now is not possible to set the new reference synchronously (nor before changing control mode).

I see two solutions:

1. Make this a warning to allow users to set a reference before changing control mode:

    https://github.com/robotology/gym-ignition/blob/551bd20424bbf38db8117d7e402a5dbcd31ebe35/ignition/src/IgnitionRobot.cpp#L483-L487

1. Get the current joint positions before switching to `Position` mode and use them as a reference. In this way, if there's a small delay between the switch of the control mode and the setting of the first reference, the PIDs are already in working state.

Do you have any comment on these two, or a third preferable option?

Reply:
> Get the current joint positions before switching to Position mode and use them as a reference. In this way, if there's a small delay between the switch of the control mode and the setting of the first reference, the PIDs are already in working state.

In some cases, this is not going to work (for example, if you do that on a KUKA robot with RSI, the controller will fault due to the difference between the current reference and the current position, that are different due to static error). 

> Do you have any comment on these two, or a third preferable option?

To be honest, I am not enthusiastic of either solution (mainly because it is  not clear for the user) for me, it is ok to just take note of this problem, and understand how to solve it in the future.

Reply:
> In some cases, this is not going to work (for example, if you do that on a KUKA robot with RSI, the controller will fault due to the difference between the current reference and the current position, that are different due to static error).

I didn't get something. How could the user provide a reference more precise than the current state read by sensors? In any case there will be a mismatch, right?

Reply:
Opened a new issue to track this https://github.com/robotology/gym-ignition/issues/107

Reply:
> I didn't get something. How could the user provide a reference more precise than the current state read by sensors? In any case there will be a mismatch, right?

We can discuss f2f, but basically in general the reference is different from the measured one if there is static error. If the static error is not negligiible, switching the reference from the old reference to the new reference creates a discontinuity.